lib/timers.js

```diff
@@ -62,55 +118,80 @@ function listOnTimeout() {
   var now = Timer.now();
   debug('now: %s', now);
 
-  var diff, first, threw;
-  while (first = L.peek(list)) {
-    diff = now - first._idleStart;
+  var diff, timer, threw;
+  while (timer = L.peek(list)) {
+    diff = now - timer._idleStart;
+
+    // Check if this loop iteration is too early for the timer.
+    // This happens if one of two things is true:
+    // - There are more timers scheduled for later in the list. (Most common)
+    // - The earliest timer was canceled (unenrolled). (Less common)
     if (diff < msecs) {
       list.start(msecs - diff, 0);
       debug('%d list wait because diff is %d', msecs, diff);
       return;
-    } else {
-      L.remove(first);
-      assert(first !== L.peek(list));
+    }
 
-      if (!first._onTimeout) continue;
+    // The actual logic for when a timeout happens.
 
-      // v0.4 compatibility: if the timer callback throws and the
+    L.remove(timer);
+    assert(timer !== L.peek(list));
+
+    if (!timer._onTimeout) continue;
+
+    var domain = timer.domain;
+    if (domain) {
+
+      // If the timer callback throws and the
       // domain or uncaughtException handler ignore the exception,
       // other timers that expire on this tick should still run.
       //
       // https://github.com/joyent/node/issues/2631
-      var domain = first.domain;
-      if (domain && domain._disposed)
-        continue;
```

Comment:
@misterdjules i changed this back to `continue` as it was in the regular timers. In `_unrefActive()` it was `return`. This still seems like it could be applicable but we don't have a test for it. Could you clarify and/or help make a test if one is necessary?